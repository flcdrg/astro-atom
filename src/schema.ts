import { z } from 'astro/zod';

export const atomSchema = z.object({
	title: z.string().optional(),
	description: z.string().optional(),
	updated: z.string(),
	customData: z.string().optional(),
	categories: z.array(z.string()).optional(),
	author: z.string().optional(),
	commentsUrl: z.string().optional(),
	source: z.object({ url: z.string().url(), title: z.string() }).optional(),
	enclosure: z
		.object({
			url: z.string(),
			length: z.number().nonnegative().int().finite(),
			type: z.string(),
		})
		.optional(),
	link: z.string().optional(),
	content: z.string().optional(),
});
