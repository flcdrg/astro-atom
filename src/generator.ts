/**
 * Default generator information for Atom feeds
 */
export const defaultGenerator = {
  uri: 'https://github.com/flcdrg/astrojs-atom',
  value: 'astrojs-atom',
  version: '1.0.0' // This will be updated by the update-version.js script
};