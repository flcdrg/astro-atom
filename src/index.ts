import { z } from 'astro/zod';
import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import { atomSchema, atomEntrySchema } from './schema.js';
import { errorMap } from './util.js';

export { atomSchema };

type UnwrapArray<T> = T extends Array<infer U> ? U : never;

export type AtomPerson = UnwrapArray<z.infer<typeof atomSchema>['author']>;
export type AtomLink = UnwrapArray<z.infer<typeof atomSchema>['link']>;
export type AtomCategory = UnwrapArray<z.infer<typeof atomSchema>['category']>;
export type AtomContributor = UnwrapArray<z.infer<typeof atomSchema>['contributor']>;
export type AtomGenerator = z.infer<typeof atomSchema>['generator'];
export type AtomSource = z.infer<typeof atomEntrySchema>['source'];
export type AtomEntry = z.infer<typeof atomEntrySchema>;

export type AtomFeedOptions = z.infer<typeof atomSchema> & {
  stylesheet?: string | boolean;
  xmlns?: Record<string, string>;
  lang?: string; // Add lang option to specify xml:lang attribute
};

export default async function getAtomResponse(atomOptions: AtomFeedOptions): Promise<Response> {
  const atomString = await getAtomString(atomOptions);
  return new Response(atomString, {
    headers: {
      'Content-Type': 'application/xml',
    },
  });
}

export async function getAtomString(atomOptions: AtomFeedOptions): Promise<string> {
  // Extract non-schema options
  const { stylesheet, xmlns, lang, ...schemaOptions } = atomOptions;
  const validated = await validateAtomOptions(schemaOptions);
  // Pass through stylesheet, xmlns and lang for XML generation
  return await generateAtom({ ...validated, stylesheet, xmlns, lang });
}

async function validateAtomOptions(atomOptions: z.infer<typeof atomSchema>) {
  const parsedResult = await atomSchema.safeParseAsync(atomOptions, { errorMap });
  if (parsedResult.success) {
    return parsedResult.data;
  }
  const formattedError = new Error(
    [
      `[Atom] Invalid or missing options:`,
      ...parsedResult.error.errors.map((zodError) => {
        const path = zodError.path.join('.');
        const message = `${zodError.message} (${path})`;
        return message;
      }),
    ].join('\n'),
  );
  throw formattedError;
}

/** Generate Atom 1.0 feed */
async function generateAtom(atomOptions: z.infer<typeof atomSchema> & { 
  stylesheet?: string | boolean; 
  xmlns?: Record<string, string>;
  lang?: string; // Add lang parameter
}): Promise<string> {
  const xmlOptions = {
    ignoreAttributes: false,
    suppressEmptyNode: true,
    suppressBooleanAttributes: false,
    format: true,
  };
  const parser = new XMLParser(xmlOptions);
  const root: any = { '?xml': { '@_version': '1.0', '@_encoding': 'UTF-8' } };
  if (typeof atomOptions.stylesheet === 'string') {
    const isXSL = /.xsl$/i.test(atomOptions.stylesheet);
    root['?xml-stylesheet'] = {
      '@_href': atomOptions.stylesheet,
      ...(isXSL && { '@_type': 'text/xsl' }),
    };
  }

  // Atom root element and namespace
  root.feed = { '@_xmlns': 'http://www.w3.org/2005/Atom' };
  
  // Add xml:lang if specified
  if (atomOptions.lang) {
    root.feed['@_xml:lang'] = atomOptions.lang;
  }

  // Additional namespaces
  if (atomOptions.xmlns) {
    for (const [k, v] of Object.entries(atomOptions.xmlns)) {
      root.feed[`@_xmlns:${k}`] = v;
    }
  }

  // Feed metadata
  root.feed.id = atomOptions.id;
  root.feed.title = atomOptions.title;
  root.feed.updated = atomOptions.updated;
  if (atomOptions.subtitle) root.feed.subtitle = atomOptions.subtitle;
  if (atomOptions.rights) root.feed.rights = atomOptions.rights;
  if (atomOptions.icon) root.feed.icon = atomOptions.icon;
  if (atomOptions.logo) root.feed.logo = atomOptions.logo;
  if (atomOptions.generator) root.feed.generator = atomOptions.generator;

  if (atomOptions.author) root.feed.author = atomOptions.author;
  if (atomOptions.link) {
    root.feed.link = atomOptions.link.map((l) =>
      Object.fromEntries(
        Object.entries(l).map(([k, v]) => ["@_" + k, v])
      )
    );
  }
  if (atomOptions.category) root.feed.category = atomOptions.category;
  if (atomOptions.contributor) root.feed.contributor = atomOptions.contributor;

  // Custom data (inserted as raw XML)
  if (typeof atomOptions.customData === 'string') {
    Object.assign(
      root.feed,
      parser.parse(`<feed>${atomOptions.customData}</feed>`).feed,
    );
  }

  // Entries
  root.feed.entry = atomOptions.entry
    .sort((a, b) => {
      const aUpdated = new Date(a.updated).getTime();
      const bUpdated = new Date(b.updated).getTime();
      return bUpdated - aUpdated;
    })
    .map((entry) => {
    const e: Record<string, unknown> = {
      id: entry.id,
      title: entry.title,
      updated: entry.updated,
    };
    if (entry.author) e.author = entry.author;
    if (entry.link) {
      e.link = entry.link.map((l) =>
        Object.fromEntries(
          Object.entries(l).map(([k, v]) => ["@_" + k, v])
        )
      );
    }
    if (entry.category) e.category = entry.category;
    if (entry.contributor) e.contributor = entry.contributor;
    if (entry.published) e.published = entry.published;
    if (entry.rights) e.rights = entry.rights;
    if (entry.source) e.source = entry.source;
    if (entry.summary) e.summary = entry.summary;
    if (entry.content) e.content = entry.content;
    if (typeof entry.customData === 'string') {
      Object.assign(e, parser.parse(`<entry>${entry.customData}</entry>`).entry);
    }
    return e;
  });

  return new XMLBuilder(xmlOptions).build(root);
}
