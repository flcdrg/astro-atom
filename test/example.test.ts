import { get } from "http";
import { AtomFeedItem, atomSchema, getAtomString } from "../src/index";
import { expect, test } from "vitest";

export function sum(a, b) {
    return a + b
  }

test('adds 1 + 2 to equal 3', () => {
    expect(sum(1, 2)).toBe(3)
  });

test('can parse item', () => {
    const atomItem = {
        title: "Test Title",
        description: "Test Description",
        updated: new Date().toISOString(),
        categories: ["test", "example"],

    };
    const result = atomSchema.safeParse(atomItem);
    expect(result.success).toBe(true);
});

test('generates valid Atom feed', async () => {
    const result = await getAtomString({
        title: "Test Feed",
        description: "Test Description",
        site: "https://example.com",
        items: [
            {
                title: "Test Item",
                description: "Test Description",
                updated: new Date().toISOString(),
                categories: ["test", "example"],
            },
        ]});
    
    expect(result).toMatchFileSnapshot("testFeed.xml");
});