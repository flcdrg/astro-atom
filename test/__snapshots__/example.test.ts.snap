// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates valid Atom feed 1`] = `"<?xml version="1.0" encoding="UTF-8"?><feed xmlns="http://www.w3.org/2005/Atom"><title>Test Feed</title><subtitle>Test Description</subtitle><id>https://example.com/</id><link href="https://example.com/"/><updated>2025-04-18T10:32:38.424Z</updated><entry><title>Test Item</title><id>https://example.com/#entry-wdw4bn2hsc</id><updated>2025-04-18T10:32:38.424Z</updated><summary>Test Description</summary><category term="test"/><category term="example"/></entry></feed>"`;
